text/0558-edition-detection.md

```diff
@@ -0,0 +1,136 @@
+- Start Date: (2019-11-20)
+- Relevant Team(s): (Ember.js, Ember CLI, Ember Data)
+- RFC PR: https://github.com/emberjs/rfcs/pull/558
+- Tracking: (leave this empty)
+
+# Edition detection
+
+## Summary
+
+Introduces a mechanism that an application can use to specify which Edition of
+Ember it is intending to target. This RFC will define:
+
+* How to specify the edition that the application is using
+* How other packages (addons, codemods, etc) can detect the applications intended edition
+* What the edition should be used for
+
+## Motivation
+
+As Ember approaches it's first edition (see
+[emberjs/rfcs#364](https://github.com/emberjs/rfcs/pull/364)) various addons
+need to modify their behavior based on the edition that is being used. An
+initial implementation (done without RFC) used the `setEdition` method from
+`@ember/edition-utils` inside the application's (or addon's) `.ember-cli.js`
+file to specify which edition was to be used. That implementation worked well
+enough throughout the intial preview period, but a number of major issues were
+(rightfully!) surfaced by the community:
+
+1. it seems unnecessary/redundant
+2. it's not clear what this flag actually does (likely due to having no RFC!)
+3. it's not statically analyzable (and therefore cannot be used by things like codemods)
+
+This RFC will review these issues in light of the updated implementation,
+showing how each of the concerns have been met.
+
+## Detailed design
+
+### Specifying the edition
+
+A new entry will be added to the project's `package.json`:
+
+```js
+{
+  "name": "project-name",
+  // ...snip
+  "ember": {
+    "edition": "octane"
+  }
+}
+```
+
+The value will be expected to be one of the valid edition names (currently
+`classic` and `octane`). Using the `package.json` for this allows us to ensure
+that the value is statically analyzable and easy to discover.
+
+### Valid use of the edition value
+
+The edition flag should only be used by addons to determine what blueprint
+output to generate and to provide helpful warnings (or errors) at build time.
```

Comment:
What about changing the build? For instance loading different class implementations. 

What about feature "detection"? Should we use https://github.com/pzuraq/ember-compatibility-helpers instead?

Reply:
Yes. Feature detection (providing alternate implementations) is generally **not** about what edition the host app is using, but instead about feature sets of a given package (most often Ember, but also applies to ember-data, or any other package). 

Using ember-compatibility-helpers’s `gte` is the right tool for that job.

Reply:
edition really is specifying a minimum version and set of features, you'll still want to use feature detection to go beyond that. You can mix: user-has-specified-octane and user-has-feature using a combo of compatibility helpers and debug-macros+this property.

Reply:
Updated prose to state the official position explicitly, thanks for surfacing the question!